### PR TITLE
EmbedSOM+optimizations

### PR DESCRIPTION
**Pull request by exaexa**
EmbedSOM+optimizationsSummary (commit messages):

```
commit 082a9eff5bd42070448d38ae67aa428e21ec3124

    High-level performance optimization
    
    Mainly including:
    - use of NearestNeighbors trees for finding the BMU
    - abuse linearity to change the neighborhood update calculation from
      expand-then-sum to more efficient sum-then-expand

commit b280b87f7d61f5cbc71bb703d4c86fbc778fd02b

    Implement EmbedSOM dimensionality reduction
```

Note: I rebased my changes atop the new `getRadius` functionality and it seems compatible, but please check that I merged it correctly and everything works as intended.

Thanks!
-mk

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -12,6 +12,7 @@ Distributions = "31c24e10-a181-5473-b8eb-7969acd0382f"
 FCSFiles = "d76558cf-badf-52d4-a17e-381ab0b0d937"
```
Comment (laurentheirendt) [inline Project.toml:12]: Why might not need `Distributions` anymore.

File: src/core.jl [hunk 2]
```diff
@@ -61,66 +61,69 @@ end
...
+                      rStart = 0.0, decay = "linear", rFinal=0.1,
+                      epochs = 10)
 
     train = convertTrainingData(train)
 
     # set default radius:
+
     if rStart == 0.0
-        rStart = √(som.xdim^2 + som.ydim^2) / 2
+        rStart = √(som.xdim^2 + som.ydim^2) / 3
```
Comment (laurentheirendt) [inline src/core.jl]: Why `/ 3`?

File: src/core.jl [hunk 3]
```diff
@@ -130,46 +133,30 @@ end
...
 vectors and the adjustment in radius after each epoch.
 
 # Arguments:
 - `x`: training Data
 - `codes`: Codebook
-- `dm`: distance matrix of all neurons of the SOM
-- `kernelFun`: distance kernel function of type fun(x, r)
-- `r`: training radius
-- `toroidal`: if true, the SOM is toroidal.
+- `nnTree`: knn-compatible tree built upon the codes
```
Comment (laurentheirendt) [inline src/core.jl]: Maybe `tree` is a better variable name.

File: src/core.jl [hunk 4]
```diff
@@ -236,15 +223,23 @@ Return a new neighbourhood radius
...
-		timeConstant = (initRadius - 1.0) / epochs
-		return initRadius - (iteration * timeConstant)
-	elseif decay == "exp"
-		timeConstant = epochs / log(initRadius)
-		return initRadius * exp(-iteration / timeConstant)
-	end
+function getRadius(initRadius::Float64, finalRadius::Float64, iteration::Int64, decay::String, epochs::Int64)
+
+    # prevent various mathematical trouble
+    @assert initRadius > 0
```
Comment (laurentheirendt) [inline src/core.jl]: Within functions, it is better to use true if conditions and customize the error message.

File: src/core.jl [hunk 5]
```diff
@@ -236,15 +223,23 @@ Return a new neighbourhood radius
...
+    @assert finalRadius > 0
+
+    # get the scaled time point in training (from 0.0 to 1.0)
+    if epochs>1
+        epochs -= 1
+    end
+
+    scaledTime = Float64(iteration-1) / Float64(epochs)
+
+    if decay == "linear"
```
Comment (laurentheirendt) [inline src/core.jl]: What if `decay` is neither `linear` nor `exp`?

File: src/satellites.jl [hunk 6]
```diff
@@ -31,16 +31,15 @@ end
...
+(The gaussian is not "normal", but that makes it compare nicely with other kernels with r used as radius.)
 
 # Arguments
 
 """
-function gaussianKernel(x::Array{Float64, 1}, r::Float64)::Array{Float64, 1}
-
-    return Distributions.pdf.(Distributions.Normal(0.0,r/3), x)
+function gaussianKernel(x, r::Float64)
+    return exp.(x.^2 .* (-1 / r^2))
```
Comment (laurentheirendt) [inline src/satellites.jl]: nice! 👍

File: src/embedding.jl [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        for i in 1:k
+            sum += kndist[i] / i
+            ssum += 1.0/i
+            if kndist[i] < min
+                min = kndist[i]
+            end
+        end
+
+        #0.000001 avoids hitting zeroes
+        sum = -ssum / (0.00000001 + sum * boost)
```
Comment (laurentheirendt) [inline src/embedding.jl]: Using scientific notation might be better here.

File: src/embedding.jl [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        dRes = [ (@spawnat w embedGigaSOM_internal(som, localpart(dData), t, k, adjust, boost)) for w in workers() ]
+
+        #hopefully the data are separated to localparts in correct order...
+        return vcat([fetch(r) for r in dRes]...)
+    else
+        return embedGigaSOM_internal(som, data, t, k, adjust, boost)
+    end
+end
+
+function embedGigaSOM_internal(som::GigaSOM.Som, data::Array{Float64,2}, t, k, adjust, boost)
```
Comment (laurentheirendt) [inline src/embedding.jl]: Is there a function documentation somewhere?

File: src/embedding.jl [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+Example:
+
+Produce a 2-column matrix with 2D cell coordinates:
+
+```
+e = embedGigaSOM(som, data)
+```
+
+Plotting is best done using histograms; e.g. with Gadfly:
```
Comment (laurentheirendt) [inline src/embedding.jl]: Is this description of plotting related to `embedGigaSOM`? Maybe this would fit better in a tutorials and or the visualization folder

File: src/embedding.jl [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+```
+draw(PNG("output.png",20cm,20cm),
+     plot(x=e[:,1], y=e[:,2],
+     Geom.histogram2d(xbincount=200, ybincount=200)))
+```
+
+Data must have the same number of dimensions as the training dataset
+and will be normalised with the same parameters.
+"""
+function embedGigaSOM(som::GigaSOM.Som, data::DataFrame;
```
Comment (laurentheirendt) [inline src/embedding.jl:36]: Can you add a few more comments to this function? Hard to follow ...

File: src/embedding.jl [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    t = knnTreeFun(Array{Float64,2}(transpose(som.codes)))
+
+    ndata = size(data, 1)
+    dim = size(data, 2)
+
+    nWorkers = nprocs()
+    if nWorkers > 1
+        dData = distribute(data)
+
+        dRes = [ (@spawnat w embedGigaSOM_internal(som, localpart(dData), t, k, adjust, boost)) for w in workers() ]
```
Comment (laurentheirendt) [inline src/embedding.jl]: I think `enumerate(workers)` would be safer.

File: src/embedding.jl [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+        #hopefully the data are separated to localparts in correct order...
+        return vcat([fetch(r) for r in dRes]...)
```
Comment (laurentheirendt) [inline src/embedding.jl:71]: nice! 👍 The order will be ok if the workers have been enumerated (see comment above)

File: src/embedding.jl [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+        mtx = zeros(Float64, 2, 3)
+
+        for i in 1:k
+            idx = knidx[i]
+            ix = Float64(som.grid[idx,1])
+            iy = Float64(som.grid[idx,2])
+            ip = kndist[i]
+
+            # a bit of gravity to avoid singularities
+            gs = 0.00001 * ip
```
Comment (laurentheirendt) [inline src/embedding.jl]: Scientific notation

File: src/embedding.jl [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+
+            mtx[1,1] += gs
```
Comment (laurentheirendt) [inline src/embedding.jl:143]: I would need a few more comments and/or more self-explaining variable names to follow properly here.

**Review by laurentheirendt — CHANGES_REQUESTED** [review]

File: src/embedding.jl (hunk 12 above)
Comment (exaexa) [reply src/embedding.jl:71]: Good to know, thanks a lot!

**Comment by exaexa** [comment]
@laurentheirendt I have hopefully fixed most of the issues you raised, thanks again for review.

Additionally, you commented on the `getRadius` function internals that @oHunewald added into `develop`; which I rewrote later anyway (in PR #55, now closed). For this reason I've moved this rewrite here as well, since there's probably no reason in postponing the fixes. All related commits should be nicely separated from the rest.

Also, since PR #56 is comparatively tiny and just follows the pattern introduced in the other changes here, I merged that as well (that should be the last commit in the series now).

Sorry for expanding the PR, but since all the changes are very related it's probably easier&more meaningful this way.

**Review by oHunewald — APPROVED** [review]

File: src/satellites.jl [hunk 15]
```diff
@@ -31,16 +31,15 @@ end
...
+Return Gaussian(x) for μ=0.0 and σ = r.
+
+(The gaussian is not "normal", but that makes it compare nicely with other kernels with r used as radius.)
 
 # Arguments
 
 """
-function gaussianKernel(x::Array{Float64, 1}, r::Float64)::Array{Float64, 1}
-
-    return Distributions.pdf.(Distributions.Normal(0.0,r/3), x)
```
Comment (oHunewald) [inline src/satellites.jl:43]: by removing Distributions.pdf I loose sensitivity. I don't see a significant performance issue. Run time is almost identical

**Review by oHunewald — CHANGES_REQUESTED** [review]

File: src/core.jl [hunk 16]
```diff
@@ -188,15 +189,17 @@ function mapToGigaSOM(som::Som, data::DataFrame)
...
 
     if nWorkers > 1
         # distribution across workers
+        dData = distribute(data)
+
         R = Array{Future}(undef,nWorkers, 1)
          @sync for (p, pid) in enumerate(workers())
              @async R[p] = @spawnat pid begin
-                 visual(som.codes, localpart(dData))
+                 vcat(knn(nnTree, transpose(localpart(dData)), 1)[1]...)
```
Comment (oHunewald) [inline src/core.jl]: just a short comment what is done here

File: src/GigaSOM.jl [hunk 17]
```diff
@@ -11,14 +11,15 @@ module GigaSOM
     using DataFrames
     using Distances
     using Distributed
-    using Distributions
```
Comment (oHunewald) [inline src/GigaSOM.jl]: I'd like to keep the original kernel function

**Review by oHunewald — CHANGES_REQUESTED** [review]

File: src/core.jl (hunk 2 above)
Comment (oHunewald) [reply src/core.jl]: / 2 was from the original som implementation, in my view it is a good starting radius if not specified otherwise

**Review by oHunewald — APPROVED** [review]

**Comment by laurentheirendt** [comment]
The tests are not passing yet ...

**Comment by exaexa** [comment]
Oh, it seems that I neglected one change in the code that is actually breaking the new way of computing gaussians -- the `gaussianKernel` is strictly typed to process 1D float arrays, but is in fact generic even to 2D arrays. Simplest fix is to remove the type annotations (I'll push that ASAP), but I don't know if that is desirable.

Is there any way to say "Any Float array type" in julia, so that the typechecking stays there?

Thanks!

**Comment by laurentheirendt** [comment]
@exaexa, just leave the strict typing out for now ...

File: src/core.jl (hunk 2 above)
Comment (laurentheirendt) [reply src/core.jl]: @oHunewald, so is `/2` more appropriate than `/3`?

**Comment by oHunewald** [comment]
I have always tested by starting radius

rStart = √(som.xdim^2 + som.ydim^2) / 2

I don't see a reason to change that. Otherwise one can pass a starting
radius as paramter

Am Di., 3. Sept. 2019 um 10:54 Uhr schrieb Laurent Heirendt <
notifications@github.com>:

File: src/core.jl (hunk 2 above)
Comment (oHunewald) [reply src/core.jl]: I would like to keep /2 as this has been our default radius all the time and gives reasonable results. Otherwise one can pass different start radius as parameter

**Comment by laurentheirendt** [comment]
may you change that? thanks.

**Comment by exaexa** [comment]
> may you change that? thanks.

It is already removed, reverted it a few commits back when reverting the gaussians.

**Review by laurentheirendt — APPROVED** [review]

**Comment by laurentheirendt** [comment]
Thanks @exaexa and @oHunewald. Great work! 👍
